### DISCUSSION TITLE
How to access D1 database in test for cloudflare worker?

### DISCUSSION THREAD

**Discussion by stmqt2023** [vitest-dev/vitest#3986] [Q&A]
Environment:
Wrangler - 3.5.1
Miniflare - 3.20230814.1
Vitest - 0.33.0
Vitest-environment-miniflare - 2.14.0

I created a D1 database and I can successfully access it in my cloudflare worker. I have this added in my "wrangler.toml" file:

```
[[d1_databases]]
binding = "DB" # i.e. available in your Worker on env.DB
database_name = "edge_db"
database_id = "<id>"
```

Then in the worker code (worker.ts), I can run SQL queries by using "env.DB" (calling env.DB.prepare(), then run() on the resulting statement. It finds the tables I have in the database and can INSERT / SELECT to/from the tables.

I also have a vitest unit test that tests the worker.  I can run the worker from it successfully.  How do I access the database from this test?

```
import {afterAll, beforeAll, describe, expect, it} from "vitest";

const {unstable_dev} = require("wrangler");

describe("Test", () => {
	let worker;

	beforeAll(async () => {
		worker = await unstable_dev("src/worker.ts", {
			experimental: {disableExperimentalWarning: true},
		});
	});

	afterAll(async () => {
		await worker.stop();
	});

	it("test worker.", async () => {
		const login_page = "<login page>";
		const resp = await worker.fetch(login_page);

		if (resp) {
			const text = await resp.text();
			expect(text).contains("some text");
		}
	});
});
```

My vitest.config.ts is as follows:

```
import {defineConfig} from "vitest/config";
import {Log, LogLevel} from "miniflare";

export default defineConfig({
    test: {
        environment: "miniflare",
        // Configuration is automatically loaded from `.env`, `package.json` and
        // `wrangler.toml` files by default, but you can pass any additional Miniflare
        // API options here:
        environmentOptions: {
            bindings: {KEY: "value"},
            kvNamespaces: ["TEST_NAMESPACE"],
						log: new Log(LogLevel.DEBUG)
        },
    },
})
```

The way I am running the test is using this command from the terminal:

`NODE_OPTIONS=--experimental-vm-modules npx vitest run`

I have tried the following with no success:

- Rename the binding to __D1_BETA__DB
- Use getMiniflareBindings() to get the environment or the DB (named either "DB" or "__D1_BETA__DB". e.g.

`const env = getMiniflareBindings(); `

or

`const { __D1_BETA__DB } = getMiniflareBindings()`

When I try to run a statement using env.__D1_BETA__DB or __D1_BETA__DB in the unit test, it cannot find the tables that I have in my database.   What is the correct way to be able to access the database in the unit test? 

Thanks.

**Comment by pop-srw** [comment]
I guess you have to use pre-release version of vitest-pool-workers lib to internally use miniflare to run the test and access to D1 test database.

You can follow this step.
1. Install pre-release vitest-pool-worker.
`npm install -D https://prerelease-registry.devprod.cloudflare.dev/workers-sdk/runs/8161335149/npm-package-cloudflare-vitest-pool-workers-5070`

2. set vitest config to use wrangler.toml config.
```
// vitest.config.js
import { defineWorkersConfig } from '@cloudflare/vitest-pool-workers/config'

export default defineWorkersConfig({
  test: {
    poolOptions: {
      workers: {
        wrangler: { configPath: './wrangler.toml' },
      },
    },
  },
})
```

3. define an env type to match wrangler config.
```
// env.d.ts
declare module 'cloudflare:test' {
  interface ProvidedEnv {
    D1: D1Database
  }
}
```

4. If you want to access d1 directly use can write a test.
```
// xxx.test.ts
import { env } from 'cloudflare:test'

describe('authentication integration tests', () => {
  beforeAll(async () => {
    await env.D1.prepare('[sql to create tables]').run()
  })
  
  test(async () => {
    const results = await env.D1.prepare('[sql to test]').all()
    [test whatever]
  }
})
``` 

5 But If you want to mock D1 database for endpoint request. You have to create a mock env and pass env.D1 to the mock env. I'm using hono framework, I'm not sure how to implement it in plain cloudflare worker, so I write a sample code so you can get the idea. So just update the xxx.test.ts as below
```
// xxx.test.ts
import { env } from 'cloudflare:test'
// import app from '../index.ts' which is the index file for my app
import app from '..'

const MOCK_ENV = {
  D1: env.D1,
}
describe('authentication integration tests', () => {
  beforeAll(async () => {
    await env.D1.prepare('[sql to create tables]').run()
  })
  
  test(async () => {
    const res = await app.request('/', { method: 'get' }, MOCK_ENV)
    expect(res.status).toBe(200)
    [test whatever]
  }
})
```

I attached references I used
[https://github.com/cloudflare/workers-sdk/issues/5127](url)
[https://github.com/mrbbot/vitest-pool-workers-prerelease-getting-started](url)

Hope that's help.
